### ISSUE TITLE
large files use too much memory

### ISSUE THREAD

**Issue by jtmoon79** [jtmoon79/super-speedy-syslog-searcher#76]
tl;dr multiple large ~gzip~ files cause `s4` to use an enormous amount of memory

### Problem

Currently, the "streaming" mode, where `Block`s are explicitly dropped, occurs during the `SyslogProcessor` stage `Stage3StreamSyslines`.
https://github.com/jtmoon79/super-speedy-syslog-searcher/blob/0.2.48/src/bin/bin.rs#L1444-L1488

But the initial search for a matching datetime occurs during `SyslogProcessor` previous stage `Stage2FindDt`
https://github.com/jtmoon79/super-speedy-syslog-searcher/blob/0.2.48/src/bin/bin.rs#L1376-L1424

This means, all of the data read from the gzip file is stored in memory during the stage `Stage2FindDt` search for matching datetime.
https://github.com/jtmoon79/super-speedy-syslog-searcher/blob/0.2.48/src/readers/blockreader.rs#L1674-L1677
That stored data is not dropped until stage 3 `Stage3StreamSyslines` processing loops once.

In theory, this delayed drop affect only happens when the user passes datetime filters (`-a`, `-b`). In theory,wWithout datetime filters, the stage `Stage2FindDt` will only read the first block and then proceeed to stage `Stage3StreamSyslines` where explicit dropping can occur.

### Solution

gzip and xz files must explicitly drop data sooner than other files, during Stage `Stage2FindDt`.

**Comment by jtmoon79** [comment]
> In theory, this delayed drop affect only happens when the user passes datetime filters (`-a`, `-b`). In theory,wWithout datetime filters, the stage `Stage2FindDt` will only read the first block and then proceeed to stage `Stage3StreamSyslines` where explicit dropping can occur.

In practice, not passing `-a` or `-b` also causes `s4` to use a claim a large amount of memory.

**Comment by jtmoon79** [comment]
good news: this affects all files, not just compressed files.
bad news: this affects all files, not just compressed files.

😣

**Comment by jtmoon79** [comment]
Essentially, my implementation of "proactive Block dropping" is not working as expected
_or_
there are some other data structures that also grow as a file as a read, and those are not dropped.
